### PR TITLE
Limit purchasing of stock options and add progress bar to company view

### PR DESCRIPTION
**Pull request by vipulyadav150**
### #14 Solved : 
 * ### _Users can now own only limited amount of stocks from a company and the calculations are based on cap_type of _company_ as given below:_
     *Small Cap - 18%
     *Mid Cap - 12%
     *Large Cap - 8%
* ### _Also added a field max_stocks_sell in Company model to handle this_

### CODE REVIEW

File: market/models.py [hunk 1]
```diff
@@ -54,21 +53,28 @@ def get_absolute_url(self):
...
             self.save()
             return True
         return False
 
 
 def pre_save_company_receiver(sender, instance, *args, **kwargs):
+
+    # Setting the maximum stocks that a user can own for a company
+    if instance.cap_type == 'small':
+        instance.max_stocks_sell = float(instance.stocks_offered) * 0.18
```
Comment (shan18) [inline market/models.py]: Is this line of code working? As the `max_stocks_sell` is an integer field and the value being given to it is a float.

File: market/views.py [hunk 2]
```diff
@@ -64,25 +64,30 @@ def post(self, request, *args, **kwargs):
...
-                                num_stocks=quantity,
-                                price=price,
-                                mode=mode,
-                                user_net_worth=InvestmentRecord.objects.calculate_net_worth(user)
-                            )
-
-                            messages.success(request, 'Transaction Complete!')
+                    # Checking with max stocks a user can purchase for a company
+                    total_quantity = investment_obj.stocks + quantity
+                    if not total_quantity > company.max_stocks_sell:
```
Comment (shan18) [inline market/views.py]: Convert this to a simple comparison expression:  
`if total_quantity <= company.max_stocks_sell`  
This is more readable.

File: market/views.py [hunk 3]
```diff
@@ -64,25 +64,30 @@ def post(self, request, *args, **kwargs):
...
+                            else:
+                                messages.error(request, 'The company does not have that many stocks left!')
 
                         else:
-                            messages.error(request, 'The company does not have that many stocks left!')
-
+                            messages.error(request, 'You have Insufficient Balance for this transaction!')
                     else:
-                        messages.error(request, 'You have Insufficient Balance for this transaction!')
+                        messages.error(request, "You can own only " + str(company.max_stocks_sell) + "stocks from this company!")
```
Comment (shan18) [inline market/views.py]: Change the message to  
"This company allows each user to hold a maximum of " + str(company.max_stocks_sell) + " stocks"

**Review by shan18 — CHANGES_REQUESTED** [review]

File: market/models.py (hunk 1 above)
Comment (vipulyadav150) [reply market/models.py]: The casting of instance.stocks_offered to float was removed

File: market/views.py (hunk 2 above)
Comment (vipulyadav150) [reply market/views.py]: The expression was changed as directed and is working fine

File: market/views.py (hunk 3 above)
Comment (vipulyadav150) [reply market/views.py]: The error message was changed as directed

**Review by shan18 — APPROVED** [review]
Looks fine now. Can be merged.

**Pull request merged**